packages/flutter_test/lib/src/widget_tester.dart

```diff
@@ -100,38 +105,141 @@ typedef WidgetTesterCallback = Future<void> Function(WidgetTester widgetTester);
+  const TestDimension();
+
+  /// Returns an iterable of the variations that this test dimension represents.
+  ///
+  /// The variations returned should be unique so that the same variation isn't
+  /// needlessly run twice.
+  Iterable<T> get variations;
+
+  /// Returns the string that will be used to both add to the test description, and
+  /// be printed when a test fails for this variation.
+  String describeVariation(T variation);
+
+  /// A function that will be called before each variation is tested, with the
+  /// variation that will be tested.
+  ///
+  /// This function should preserve any state needed to restore the testing
+  /// environment back to its base state when [tearDown] is called.
+  Future<void> setUp(T variation);
+
+  /// A function that is guaranteed to be called after this variation is tested,
+  /// even if it throws an exception.
+  ///
+  /// Calling this function must return the testing environment back to the base
+  /// state it was in before [setUp] was called.
+  Future<void> tearDown(T variation);
+}
+
+/// The [TestDimension] that represents the "default" test that is run if no
+/// `matrix` is specified for [testWidgets].
+///
+/// This dimension can be added into a list of other test dimensions to provide
+/// a "control" test where nothing is changed from the base test environment.
+class DefaultDimension extends TestDimension<dynamic> {
+  @override
+  final Iterable<dynamic> variations = <String>['default'];
+
+  @override
+  String describeVariation(dynamic variation) => '';
+
+  @override
+  Future<void> setUp(dynamic variation) async {}
+
+  @override
+  Future<void> tearDown(dynamic variation) async {}
+}
+
+/// A [TestDimension] that runs tests with [debugDefaultTargetPlatformOverride]
+/// set to different values of [TargetPlatform].
+class TargetPlatformDimension extends TestDimension<TargetPlatform> {
+  /// Creates a [TargetPlatformDimension] that tests the given [variations].
+  TargetPlatformDimension(this.variations)
+    : assert(variations.toSet().length == variations.length,
+             'There were ${variations.length - variations.toSet().length} platforms '
+             'that were specified more than once. Each platform may only be tested once.');
+
+  // Stores the target platform that was set when setUp was called.
+  TargetPlatform _previousTargetPlatform;
+
+  @override
+  final Iterable<TargetPlatform> variations;
+
+  @override
+  String describeVariation(TargetPlatform variation) => 'TargetPlatform ${describeEnum(variation)}';
+
+  @override
+  Future<void> setUp(TargetPlatform variation) async {
+    assert(_previousTargetPlatform == null);
+    _previousTargetPlatform = debugDefaultTargetPlatformOverride;
+    debugDefaultTargetPlatformOverride = variation;
+  }
+
+  @override
+  Future<void> tearDown(TargetPlatform variation) async {
+    debugDefaultTargetPlatformOverride = _previousTargetPlatform;
+    _previousTargetPlatform = null;
+  }
```

Comment:
it's cleaner if setUp returns something that is then passed to tearDown (lets you make TargetPlatformDimension const, for instance)

Reply:
sure